src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs

```diff
@@ -4400,350 +4405,727 @@ internal static IEnumerable<CompletionResult> CompleteFilename(CompletionContext
 
-                            foreach (dynamic entry in psobjs)
-                            {
-                                providerPath = entry.ProviderPath;
-                                if (string.IsNullOrEmpty(providerPath))
-                                {
-                                    // This is unexpected. ProviderPath should never be null or an empty string
-                                    leaf = null;
-                                    break;
-                                }
+            foreach (var item in resolvedPaths)
+            {
+                var pathInfo = (PathInfo)item.BaseObject;
+                string baseTooltip = pathInfo.ProviderPath.Equals(string.Empty, StringComparison.Ordinal)
+                    ? pathInfo.Path
+                    : pathInfo.ProviderPath;
+                if (baseTooltip[^1] is not '\\' and not '/' and not ':')
+                {
+                    baseTooltip += provider.ItemSeparator;
+                }
 
-                                if (!notHiddenEntries.Contains(providerPath))
-                                {
-                                    notHiddenEntries.Add(providerPath);
-                                }
-                            }
+                _ = context.Helper.CurrentPowerShell
+                    .AddCommandWithPreferenceSetting("Microsoft.PowerShell.Management\\Get-ChildItem")
+                    .AddParameter("LiteralPath", pathInfo.Path);
 
-                            if (leaf != null)
-                            {
-                                leaf += "*";
-                                var parentPath = Path.GetDirectoryName(providerPath);
+                bool hadErrors;
+                var childItemOutput = context.Helper.ExecuteCurrentPowerShell(out _, out hadErrors);
 
-                                // ProviderPath should be absolute path for FileSystem entries
-                                if (!string.IsNullOrEmpty(parentPath))
-                                {
-                                    string[] entries = null;
-                                    try
-                                    {
-                                        entries = Directory.GetFileSystemEntries(parentPath, leaf, _enumerationOptions);
-                                    }
-                                    catch (Exception)
-                                    {
-                                    }
+                var childrenInfoTable = new Dictionary<string, bool>(childItemOutput.Count);
+                var childNameList = new List<string>(childItemOutput.Count);
 
-                                    if (entries != null)
-                                    {
-                                        hiddenFilesAreHandled = true;
+                if (hadErrors)
+                {
+                    // Get-ChildItem failed to get some items (Access denied or something)
```

Comment:
I wonder if it's better to just run `Get-ChildItem -ErrorAction Ignore` instead of running twice if it fails the first time?

Reply:
If we just ignore the errors then it will leave out some results:
![image](https://github.com/PowerShell/PowerShell/assets/42123497/4a9e8e8d-1f02-45ed-85e8-1424b8eea4ee)